url11fb0973cf_raichuman9_Swimmer526_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/fr/inria/optimization/cmaes/CMAEvolutionStrategy.java

```diff
@@ -0,0 +1,2704 @@
+
+    	arx = new double[sp.getLambda()][N];
+    	population = new double[sp.getLambda()][N];
+
+    	for (i = 0; i < sp.getLambda(); ++i) {
+    		fit.fitness[i] = new IntDouble();
+    		fit.raw[i] = new IntDouble();
+    	}
+
+    	// initialization
+    	for (i = 0; i < N; ++i) {
+    		pc[i] = 0;
+    		ps[i] = 0;
+    		for (int j = 0; j < N; ++j) {
+    			B[i][j] = 0;
+    		}
+    		for (int j = 0; j < i; ++j) {
+    			C[i][j] = 0;
+    		}
+    		B[i][i] = 1;
+    		C[i][i] = diagD[i] * diagD[i];
+    	}
+    	maxsqrtdiagC = Math.sqrt(math.max(math.diag(C)));
+    	minsqrtdiagC = Math.sqrt(math.min(math.diag(C)));
+    	countCupdatesSinceEigenupdate = 0;
+    	iniphase = false; // obsolete
+
+    	/* Some consistency check */
+    	for (i = 0; i < N; ++i) {
+    		if (LBound[i] > UBound[i])
+    			error("lower bound is greater than upper bound");
+    		if (typicalX != null) {
+    			if (LBound[i] > typicalX[i])
+    				error("lower bound '" + LBound[i] + "'is greater than typicalX" + typicalX[i]);
+    			if (UBound[i] < typicalX[i])
+    				error("upper bound '" + UBound[i] + "' is smaller than typicalX " + typicalX[i]);
+    		}
+    	}
+    	String[] s = stopConditions.getMessages();
+    	if(!s[0].equals(""))
+    		warning("termination condition satisfied at initialization: \n  " + s[0]);
+
+    	initialX = xmean.clone(); // keep finally chosen initialX
+    	
+    	timings.start = System.currentTimeMillis();
+    	timings.starteigen = System.currentTimeMillis();
+
+    	state = 0;
+    	if(options.verbosity > -1)
+    		printlnHelloWorld();
+
+    	return new double[sp.getLambda()];
+
+    } // init()
+
+    /** get default parameters in new CMAParameters instance, dimension must 
+     * have been set before calling getDefaults
+     * 
+     * @see CMAParameters#getDefaults(int)
+     */
+    public CMAParameters getParameterDefaults() {
+    	return sp.getDefaults(N);
+    }
+
+    /** get default parameters in new CMAParameters instance
+     * 
+     * @see CMAParameters#getDefaults(int)
+     */
+    public CMAParameters getParameterDefaults(int N) {
+    	return sp.getDefaults(N);
+    }
+
+    /** reads properties from default
+     * input file CMAEvolutionStrategy.properties and
+     * sets options and strategy parameter settings
+     * accordingly. Options values can be changed at any time using this function. 
+     */
+    public Properties readProperties() {
+    	return readProperties(propertiesFileName);
+    }
+    Properties properties = new Properties();
+    /** reads properties from fileName and sets strategy parameters and options
+     * accordingly
+     * @param fileName of properties file
+     */
+    public Properties readProperties(String fileName) {
+        this.propertiesFileName = fileName;
+//        if (fileName.equals(""))
+//            return properties;
+        try {
+            java.io.FileInputStream fis = new java.io.FileInputStream(fileName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fis*. The resource is referenced at line 809. The resource is closed at line 810. There are other execution paths that do not contain closure statements, for example, when Properties.load throws an exception. Ensure *fis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).